packages/core/src/routes/sign-in-experience/custom-ui-assets/index.ts

```diff
@@ -0,0 +1,105 @@
+  ...[router]: RouterInitArgs<T>
+) {
+  // TODO: Remove
+  if (!EnvSet.values.isDevFeaturesEnabled) {
+    return;
+  }
+
+  router.post(
+    '/sign-in-exp/default/custom-ui-assets',
+    koaGuard({
+      files: object({
+        file: uploadFileGuard.array().min(1),
+      }),
+      response: z.object({
+        customUiAssetId: z.string(),
+      }),
+      status: [200, 400, 500],
+    }),
+    async (ctx, next) => {
+      const { file: bodyFiles } = ctx.guard.files;
+      const file = bodyFiles[0];
+
+      assertThat(file, 'guard.invalid_input');
+      assertThat(file.size <= maxUploadFileSize, 'guard.file_size_exceeded');
+      assertThat(file.mimetype === 'application/zip', 'guard.mime_type_not_allowed');
+
+      const { experienceZipsProviderConfig } = SystemContext.shared;
+      assertThat(
+        experienceZipsProviderConfig?.provider === 'AzureStorage',
+        'storage.not_configured'
+      );
+      const { connectionString, container } = experienceZipsProviderConfig;
+
+      const { uploadFile, downloadFile, isFileExisted } = buildAzureStorage(
+        connectionString,
+        container
+      );
+
+      const [tenantId] = await getTenantId(ctx.URL);
+      assertThat(tenantId, 'guard.can_not_get_tenant_id');
+
+      const customUiAssetId = generateStandardId(8);
+      const objectKey = `${tenantId}/${customUiAssetId}/assets.zip`;
+      const errorLogObjectKey = `${tenantId}/${customUiAssetId}/error.log`;
+
+      try {
+        // Upload the zip file to `experience-zips` container, in which a blob trigger is configured,
+        // and an azure function will be executed automatically to unzip the file on blob received.
+        // If the unzipping process succeeds, the zip file will be removed and assets will be stored in
+        // `experience-blobs` container. If it fails, the error message will be written to `error.log` file.
+        await uploadFile(await readFile(file.filepath), objectKey, {
+          contentType: file.mimetype,
+        });
+
+        // Check status every 300ms.
+        await new Promise((resolve, reject) => {
+          const checkUnzipStatus = async () => {
+            const [hasZip, hasError] = await Promise.all([
+              isFileExisted(objectKey),
+              isFileExisted(errorLogObjectKey),
+            ]);
+            if (!hasZip && !hasError) {
+              resolve(true);
+            }
+            if (hasError) {
+              const errorLogFile = await downloadFile(errorLogObjectKey);
+              // Log the detailed error message
+              const errorLogContent = errorLogFile.readableStreamBody?.read().toString();
+              getConsoleLogFromContext(ctx).error(errorLogContent);
+              // Throw the detailed error message to the outer catch block
+              reject(new Error(errorLogContent ?? 'Unzipping failed.'));
+            }
+            setTimeout(checkUnzipStatus, checkStatusTimeout);
```

Comment:
this can trigger every 300ms?

Reply:
Sorry I was in a rush writting the code. Now fixed and tested thoroughly on my local machine.

Reply:
Updated with `pRetry`